urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/httpd/resource/ProcessableResource.java

```diff
@@ -0,0 +1,382 @@
+		 && this.bufferedResource.isOpen() 
+		 );
+    }
+
+
+    /**
+     * This method returns true if the underlying resource is read-only (in general).
+     *
+     * @throws IOException If any IO error occurs.
+     **/
+    public  boolean isReadOnly()
+	throws IOException {
+
+	// Processable resources are always read-only.
+	return true;
+    }
+
+    /**
+     * This method returns the *actual* length of the underlying resource. This length will
+     * be used in the HTTP header fields to specify the transaction length.
+     *
+     * During read-process (you used the locks, didn't you?) the length MUST NOT change.
+     *
+     * @return the length of the resource's data in bytes.
+     * @throws IOException If any IO error occurs.
+     **/
+    public long getLength()
+	throws IOException {
+
+	if( !this.isOpen() )
+	    throw new IOException( "Cannot get the resource's data length if it was not opened." );
+
+	return this.bufferedResource.getLength();
+    }
+
+
+    /**
+     * Get the output stream to this resource.
+     *
+     * @throws ReadOnlyException If this resource was opened with the read-only flag set.
+     * @throws IOException If any other IO error occurs.
+     **/
+    public OutputStream getOutputStream()
+	throws ReadOnlyException,
+	       IOException {
+
+	throw new ReadOnlyException( "Processable resources are read-only. They have no output stream!" );
+    }
+
+    /**
+     * Get the input stream from this resource.
+     *
+     * @throws IOException If any IO error occurs.
+     **/
+    public InputStream getInputStream()
+	throws IOException {
+
+	if( !this.isOpen() )
+	    throw new IOException( "Cannot get the resource's input stream if it was not opened." );	
+
+	return this.bufferedResource.getInputStream();
+    }
+
+
+    /**
+     * Closes this resource.
+     *
+     * @return false if the resource was already closed, true otherwise.
+     **/
+    public boolean close()
+	throws IOException {
+
+	if( !this.isOpen() )
+	    return false; // Already closed (or not yet opened)
+
+
+	this.bufferedResource.close();
+
+	return true;
+    }
+    //--- END ------------------------------ AbstractResource implementation ------------------------------
+
+
+    /**
+     * For testing only.
+     **/
+    public static void main( String[] argv ) {
+
+	ProcessableResource pr = null;
+	try {
+
+	    java.util.List<String> command = new java.util.LinkedList<String>();
+	    // This will print the directory contents of cwd (Linux only)
+	    //command.add( "ls" );
+	    //command.add( "-l" );
+
+	    // This will call the PHP interpreter to run my test PHP file.
+	    // Note that 'php'/php-cli' does NOT print the HTTP response headers!
+	    // But 'php-cgi' does :)
+	    command.add( "php-cgi" );
+	    command.add( "document_root_alpha/php_test/headers.php" );
+	    ProcessBuilder pb = new ProcessBuilder( command );
+
+	    pr = new ProcessableResource( null,  // no http handler
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *pr*. The resource is referenced at lines 355, 358, 360, 375, 376. The resource is closed at line 365. There are other execution paths that do not contain closure statements, for example, when ProcessableResource.open throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).